Assets/MixedRealityToolkit.Extensions/Socketer/Scripts/TCPConnectionManager.cs

```diff
@@ -0,0 +1,228 @@
+        private readonly ConcurrentDictionary<int, SocketEndpoint> serverConnections = new ConcurrentDictionary<int, SocketEndpoint>();
+        private SocketEndpoint clientConnection;
+        private SocketerClient client;
+        private ConcurrentQueue<IncomingMessage> inputMessageQueue = new ConcurrentQueue<IncomingMessage>();
+
+        private SocketerClient server;
+        private TimeSpan timeoutInterval;
+
+        /// <summary>
+        /// Returns true if any server or client connections exist, otherwise false
+        /// </summary>
+        public bool HasConnections => (serverConnections.Count > 0 || clientConnection != null);
+
+        /// <summary>
+        /// Returns true if a connection is being attempted, otherwise false
+        /// </summary>
+        public bool IsConnecting => client != null;
+
+        /// <summary>
+        /// Returns the number of bytes currently queued for the socketer server
+        /// </summary>
+        public int OutputBytesQueued => SocketerClient.OutputQueueLength;
+
+        public TCPConnectionManager(TimeSpan timeoutInterval)
+        {
+            this.timeoutInterval = timeoutInterval;
+        }
+
+        /// <summary>
+        /// Call to begin acting as a server listening on the provided port
+        /// </summary>
+        /// <param name="port">port to listen on</param>
+        public void StartListening(int port)
+        {
+            if (server == null)
+            {
+                server = DoStartListening(port);
+            }
+        }
+
+        private SocketerClient DoStartListening(int port)
+        {
+            Debug.Log("Listening on port " + port);
+            SocketerClient newServer = SocketerClient.CreateListener(SocketerClient.Protocol.TCP, port);
+            newServer.Connected += OnServerConnected;
+            newServer.Disconnected += OnServerDisconnected;
+            newServer.Start();
+            return newServer;
+        }
+
+        /// <summary>
+        /// Call to stop acting as a server
+        /// </summary>
+        public void StopListening()
+        {
+            DoStopListening(ref server);
+        }
+
+        protected void DoStopListening(ref SocketerClient listener)
+        {
+            if (listener != null)
+            {
+                Debug.Log("Stopped listening on port " + listener.Port);
+                listener.Stop();
+                listener = null;
+            }
+        }
+
+        /// <summary>
+        /// Call to start acting as a client connected to the provided server and port
+        /// </summary>
+        /// <param name="serverAddress">server to connect to</param>
+        /// <param name="port">port to use for communication</param>
+        public void ConnectTo(string serverAddress, int port)
+        {
+            Debug.LogFormat($"Connecting to {serverAddress}:{port}");
+            client = SocketerClient.CreateSender(SocketerClient.Protocol.TCP, serverAddress, port);
```

Comment:
It would be nice to support other protocols than TCP, since the system does support it.

Though I understand Socketer may not have a connection handling system for non-TCP communication.

Reply:
socketer supports UDP as well as tcp. We could make this class more generic to support both networking protocols. I'll take another look at this.

Reply:
One thing to be careful with if you're adding UDP, is that if you're not fragmenting messages, most will get lost. You'll also need to consider flow control to avoid overloading the hardware network card. This varies according to the router you're using as well.

Reply:
for now i think we will keep this as the tcp connection manager. Enabling fragmented udp messages is probably beyond the scope of this review.